### ISSUE TITLE
Make Dispatching Synchronous

### ISSUE THREAD

**Issue by Ben-G** [Swift-Flow/Swift-Flow#11]
Currently dispatching is implemented using `dispatch_async` - that guarantees that all subscribers to the store will receive the latest state, before another state change can occur.

I however haven't encountered a lot of cases where this guarantee is necessary; usually code should not rely on receiving on intermediate state updates anyway.

The async dispatch also has a bunch of downsides - it makes the store harder to test and it makes it harder to debug the flow of actions in through the backtrace.

**In short, dispatching should become synchronous.**
